src/raw_node.rs

```diff
@@ -349,94 +379,259 @@ impl<T: Storage> RawNode<T> {
-            None => raft.raft_log.has_next_entries(),
-            Some(idx) => raft.raft_log.has_next_entries_since(idx),
-        };
-        if has_unapplied_entries {
+
+        if self.snap().map_or(false, |s| !s.is_empty()) {
             return true;
         }
-        if raft.soft_state() != self.prev_ss {
+
+        if raft
+            .raft_log
+            .has_next_entries_since(self.commit_since_index)
+        {
             return true;
         }
-        let hs = raft.hard_state();
-        if hs != HardState::default() && hs != self.prev_hs {
+
+        if !raft.msgs.is_empty() || !self.messages.is_empty() {
             return true;
         }
         false
     }
 
-    /// HasReady called when RawNode user need to check if any Ready pending.
-    /// Checking logic in this method should be consistent with Ready.containsUpdates().
-    #[inline]
-    pub fn has_ready(&self) -> bool {
-        self.has_ready_since(None)
+    fn commit_ready(&mut self, rd: Ready) {
+        if let Some(ss) = rd.ss {
+            self.prev_ss = ss;
+        }
+        if let Some(hs) = rd.hs {
+            self.prev_hs = hs;
+        }
+        let rd_record = self.records.back().unwrap();
+        assert!(rd_record.number == rd.number);
+        let raft = &mut self.raft;
+        if rd_record.has_snapshot {
+            raft.raft_log.stable_snap();
+        }
+        if rd_record.last_entry.is_some() {
+            raft.raft_log.stable_entries();
+        }
     }
 
-    /// Grabs the snapshot from the raft if available.
-    #[inline]
-    pub fn snap(&self) -> Option<&Snapshot> {
-        self.raft.snap()
+    fn commit_apply(&mut self, applied: u64) {
+        self.raft.commit_apply(applied);
     }
 
-    /// Advance notifies the RawNode that the application has applied and saved progress in the
-    /// last Ready results.
-    pub fn advance(&mut self, rd: Ready) {
-        self.advance_append(rd);
-        let commit_idx = self.prev_hs.commit;
-        if commit_idx != 0 {
-            // In most cases, prevHardSt and rd.HardState will be the same
-            // because when there are new entries to apply we just sent a
-            // HardState with an updated Commit value. However, on initial
-            // startup the two are different because we don't send a HardState
-            // until something changes, but we do send any un-applied but
-            // committed entries (and previously-committed entries may be
-            // incorporated into the snapshot, even if rd.CommittedEntries is
-            // empty). Therefore we mark all committed entries as applied
-            // whether they were included in rd.HardState or not.
-            self.advance_apply(commit_idx);
+    /// Notifies that the ready of this number has been well persisted.
+    ///
+    /// Since Ready must be persisted in order, calling this function implicitly means
+    /// all readys with numbers smaller than this one have been persisted.
```

Comment:
```suggestion
    /// Notifies that the ready of this number has been persisted.
    ///
    /// Since Ready must be persisted in order, calling this function implicitly means
    /// all readies with numbers smaller than this one have been persisted.
    ///
    /// `has_ready` and `ready` should be called later to handle further updates that become
    /// valid after ready being persisted.
```